### ISSUE TITLE
[Rule Proposal] require-array-sort-compare

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#10363]
**Please describe what the rule should do:**

This rule reports `x.sort()` calls with no arguments.

```js
//✘ BAD
xs.sort()

//✔ GOOD
xs.sort((a, b) => a - b)
ys.sort((a, b) => a.localeCompare(b))
zs.sort(cmp)
ws.sort(undefined) //← explicit default behavior
```

Because the default compare function does compare the array elements as strings then it can cause unexpected behavior. For example:

```js
[1, 2, 10].sort() //→ it becomes [1, 10, 2]
```

The language spec describes this matter.

> NOTE 2: Method calls performed by the ToString abstract operations in steps 5 and 7 have the potential to cause SortCompare to not behave as a consistent comparison function.
> https://www.ecma-international.org/ecma-262/8.0/#sec-sortcompare

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
//✘ BAD
xs.sort()

//✔ GOOD
xs.sort((a, b) => a - b)
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This is a trap that the language spec mentions.

I'm not sure if we want to check this in core rules because the check of prototype methods can make many false positive. But I thought it's worth to discuss.

**Comment by ljharb** [comment]
Why would `.sort(undefined)` be better than `.sort()`?

I'd think that if you want the explicit default behavior, you'd do `.sort((a, b) => String(b).localeCompare(a))` or similar.

**Comment by mysticatea** [comment]
Because it tells readers that "this has not forgotten to give the compare function."
I agree that `localeCompare` is a better choice.
